### ISSUE TITLE
How to build encoder.json and dict.txt

### ISSUE THREAD

**Issue by 008karan** [pytorch/fairseq#1186]
I am training RoBERTa on a different language. I found how to build vocab.bpe using other BPE methods but not able to figure out how to get dict.txt and encoder.json.
Please suggest how to do this.

**Comment by 008karan** [comment]
I am using sentencepiece BPE

**Comment by lematt1991** [comment]
You shouldn't need encoder.json at all.  Follow [these](https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.pretraining.md) instructions, and skip the "Next encode it with the GPT-2 BPE" section, and encode using your sentencepiece BPE.  The rest should be the same.

**Comment by 008karan** [comment]
But in the preprocessing step there is `srcdict` argument and train.bpe valid.bpe test.bpe files are needed whereas I only got one model file and vocab from sentencepiece BPE.
```
 fairseq-preprocess \
    --only-source \
    --srcdict gpt2_bpe/dict.txt \
    --trainpref wikitext-103-raw/wiki.train.bpe \
    --validpref wikitext-103-raw/wiki.valid.bpe \
    --testpref wikitext-103-raw/wiki.test.bpe \
    --destdir data-bin/wikitext-103 \
    --workers 60
```
@lematt1991

**Comment by lematt1991** [comment]
The *.bpe files are the names of the BPE encoded files.  You would do something like:

```BASH
for SPLIT in train valid test; do \
    cat wikitext-103-raw/wiki.${SPLIT}.raw | spm_encode --model=<model_file> --output_format=piece > wikitext-103-raw/wiki.${SPLIT}.bpe
done
```

And then:

```BASH
fairseq-preprocess \
    --only-source \
    --trainpref wikitext-103-raw/wiki.train.bpe \
    --validpref wikitext-103-raw/wiki.valid.bpe \
    --testpref wikitext-103-raw/wiki.test.bpe \
    --destdir data-bin/wikitext-103 \
    --workers 60
```

By not specifying `--srcdict` it will generate a dictionary for you.

**Comment by lematt1991** [comment]
Does this solve your problem?  If so, do you mind closing this issue.  Thanks!

**Comment by lematt1991** [comment]
Closing due to inactivity

**Comment by ceatlinar** [comment]
Hi, I am having a problem with training Roberta on Turkish. I followed your instructions and created bpe files with sentencepiece and then run fairseq-preprocess command. Everything seemed fine until I tried to get predictions from the model (the training was over without an error). I tried to make prescriptions using roberta.fill.mask, for example when I test it with roberta.fill.mask('Tugberk be Ferhat bir proje <mask> çalışıyor.') it gives an error with decoding, an int base decoding error for '_bağırma'. Can you help me with how to get predictions from the model I am trying to train for Turkish. Idk at which step there is a mistake I make

**Comment by GabboM** [comment]
what if dict.txt is not present and I have multiple data-bins (data-bin1:data-bin2:data-bin3, ec...), how can I create a general dict.txt that is valid for each data-bin and not create a new one for each of them that will also cause problems?

**Comment by Asaw2019** [comment]
@lematt1991 lematt1991 

> What BPE are you using (sentencepiece, fastbpe, something else)? The `encoder.json` is specific to [GPT2's BPE](https://github.com/openai/gpt-2/blob/master/src/encoder.py). The `dict.txt` file will get created when you preprocess your data using `fairseq-preprocess`
@lematt1991 please guide me how to create encoder.json using gpt-2 bpe for other language.

**Comment by jordiae** [comment]
> The *.bpe files are the names of the BPE encoded files. You would do something like:
> 
> ```shell
> for SPLIT in train valid test; do \
>     cat wikitext-103-raw/wiki.${SPLIT}.raw | spm_encode --model=<model_file> --output_format=piece > wikitext-103-raw/wiki.${SPLIT}.bpe
> done
> ```
> 
> And then:
> 
> ```shell
> fairseq-preprocess \
>     --only-source \
>     --trainpref wikitext-103-raw/wiki.train.bpe \
>     --validpref wikitext-103-raw/wiki.valid.bpe \
>     --testpref wikitext-103-raw/wiki.test.bpe \
>     --destdir data-bin/wikitext-103 \
>     --workers 60
> ```
> 
> By not specifying `--srcdict` it will generate a dictionary for you.

Hi @lematt1991, in case you want to use a specific dictionary, you should create by hand dict.txt, right? Afaik, fairseq-preprocess will generate the dictionary by looking at the unique tokens appearing in training, but in case some of the tokens in your original dictionary don't appear in the train set (e.g., placeholder tokens you want to reserve, or special tokens you may use in a future fine-tuning but not during pre-training), these will not be added to dict.txt, and there are no arguments to add them. In this case, one would have to manually concatenate them to dict.txt, with frequency 0?
